Added decoder to return a specific Avro record from bytes.

danharvey opened the pull request:
- Moved getByteBufferer into shared abstract class
 - Added decoder to return specific Avro records

This allows us to work with Java kafka consumers using native compiled Avro objects.

This implements #144

ewencp commented:
A couple of notes:

1. It looks like most of this patch ends up duplicating code -- only a couple of lines around the decode step are actually new. If we use this patches approach, I think it'd be better to make a shared base class to avoid a lot of the duplication.
2. I think it would be possible to just incorporate this as an alternative in the existing decoder. We could add an option for the decoder that would enable decoding specific records. Then, when we encounter a new schema, we could try to resolve the SpecificRecord type using reflection. If we find it, then we should be able to create a SpecificDatumWriter for it (one of the constructors takes a `class<? extends SpecificDataBase>`) and decode. We could cache the schema -> SpecificRecord class lookup so any subsequent messages woudn't have to do anything with reflection. I haven't written the code yet so I can't be sure, but I think this approach should work.

And thanks for working on this, I think it'll be a great addition! Working with SpecificRecords is definitely more pleasant than GenericRecords.

danharvey commented:
Thanks for looking over this.

I wasn't sure how far to go with refactoring to reuse code but I was thinking along the lines of creating a base class. Alternatively if we extended the existing decoder it wouldn't allow us to return the correct type which I think would be preferable `public T fromBytes(byte[] bytes)` which would mean creating a new generic decoder, with shared base code?

We could put a parameter on AbstractKafkaAvroDeserializer to do the switching to using the specific record then cast this in the KafkaSpecificAvroDecoder which can be generic?

A few general points, the Avro generated classes don't extend any useful interfaces for using with generics so I think allowing any class would probably make most sense.

Then to get the target schema I'm currently passing it in, but we could use some code like this (which I do in our app code) to get the schema from the Avro class. Annoyingly with generics we can't get the generic class in a reliable way to call this on but we could pass in the target avro compiled record into the constructor of the KafkaSpecificAvroDecoder for it to use.

```
    public static Schema getClassSchema(Class messageClass) {
        try {
            Method method = messageClass.getMethod("getClassSchema");
            Object o = method.invoke(null);

            return (Schema)o;
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("messageClass is not an Avro record base.");
        } catch (InvocationTargetException e) {
            throw new RuntimeException("messageClass is not an Avro record base.");
        } catch (IllegalAccessException e) {
            throw new RuntimeException("messageClass is not an Avro record base.");
        }
    }
```

Let me know if you have any thoughts, I'll look at simplifying this tomorrow.

Also do you have any license requirements for me contributing this code?

ewencp commented:
What if we implement the code in the existing deserializer using the flag, and provide a very thin wrapper that gives you the type safety if you want it? The reason I'm pushing for it working from the existing deserializer is that the new consumer supports configuring the serializer from configs instead of passing it in and I think it'd be better to be able to say you can just set `key.deserializer` and `value.deserializer` to `KafkaAvroDeserializer` and it'll work for any Avro type, you'll get back an instance of the right type. In that scenario, the type information actually doesn't accomplish anything because it's all handled at runtime and the generic types have already been erased, but the user can still specify the types for their consumer, i.e. `KafkaConsumer<String, MySpecificRecord>`.

The types *are* useful if you instantiate the decoders/deserializers yourself. But you can get it by making a class `KafkaAvroSpecificDeserializer<? extends SpecificRecordBase>` that only overrides the one method and does the cast for you. Technically we lose one compile time type check, but it's easy to convince ourselves it won't be a problem at runtime.

I think you're right that AbstractKafkaAvroDeserializer is the right place to put the flag, that's where we want to centralize all the decoding logic. I'm not sure about adding generic type parameters to `KafkAvroDecoder`. At the very least, we need to be careful that it would be a backwards compatible change.

My generated code for a sample looks like this:
```
public class User extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord
```
So wouldn't one of those work? Or do older versions of Avro not use those base classes?

And finally, the basic structure of your code for getting the schema is what I had in mind.

danharvey commented:
Updated PR. After looking over the avro code I found a cleaner way to resolve the target schema, so that made it quite simple to do this for any incoming schema as you thought.

Now it's just a single cast to go to the correct type which I think is fine.

ewencp commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -71,4 +78,28 @@ protected Object deserialize(byte[] payload) throws SerializationException {
+    }
+
+    private Schema getTargetSchema(Schema schema) {
+        Schema targetSchema = targetSchemaCache.get(schema);
```
This cache needs synchronization -- this may be accessed by multiple threads. A simple concurrent map should work fine.

danharvey replied:
Changed.

ewencp commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -71,4 +78,28 @@ protected Object deserialize(byte[] payload) throws SerializationException {
+            try {
+                targetSchema = ((SpecificRecord)targetClass.newInstance()).getSchema();
+            } catch (InstantiationException e) {
+                throw new RuntimeException("Failed to get target schema for " + targetClass, e);
```
Since this is just going to be converted to a SerializationException when it gets caught in `deserialize()`, maybe we could just throw a SerializationException directly here?

We might also want to make the error messages a bit clearer here, maybe saying that a SpecificRecord was requested by the class specified by the schema could not be found.

danharvey replied:
Makes sense, I've update the exceptions.

ewencp commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/KafkaAvroDecoder.java:
```diff
@@ -28,12 +28,17 @@ public KafkaAvroDecoder(SchemaRegistryClient schemaRegistry) {
     this.schemaRegistry = schemaRegistry;
   }
 
+    public KafkaAvroDecoder(SchemaRegistryClient schemaRegistry, boolean useSpecificDecoder) {
```
Why have this extra constructor? The one above it is just for tests and any real use will need to override the default schema registry URL so it would have to use the one that takes the properties parameter, right?

danharvey replied:
Yes that's true, I'll remove this.

danharvey replied:
I think this is useful for the tests actually. Otherwise we can't pass in a local (or mock) schemeRegistry in for the test. The alternative is we could modify the properties so it can take a schemaRegistryFactory or similar?

ewencp replied:
Ah, I see the issue. One alternative would be to still pass in the Properties object and refactor parsing of useSpecificDecoder to a separate method that can be reused.

I'm only being picky about this because this is technically a public interface even though it's just for testing and we need to be careful about compatibility. If we end up with a few more of these options, we might just end up with a bunch of different constructors to toggle them all + maintain backwards compatibility. It's definitely possible users will want to also take advantage of these constructors for their own unit testing, so getting the interface right is pretty important.

danharvey replied:
Makes sense, I've done that.

ewencp commented:
@danharvey Awesome, this looks great and is pretty simple. Two general comments:

1. We should at least have a simple test. I think it should be a pretty easy addition to KafkaAvroSerializerTest, although you'll need to instantiate two KafkaAvroDecoders in order to test both modes.
2. Code formatting -- I know it's annoying, but we want to keep this clean. We're using Google's code style, which is pretty easy to get into a bunch of common IDEs. Alternatively, I think the only important issue was some 4-space indents instead of 2-space indents.

@junrao should probably take a pass at this as well since he's more familiar with this code and this is an area where we need to vet changes very carefully.

danharvey commented:
No worry with the style, I was wondering what you used for that.

I've added a simple test. To be able to use the Decoder with the specific flag enabled we need a class that matches what the schema specifies, then this needs to behave as a SpecificRecord would and return it's schema. I've added a generated Avro file to make this simple, but not sure that's something you'll want committed?

ewencp commented:
Excellent, the test looks good to me.

On the generated Avro -- I'd recommend using avro-maven-plugin to generate it on the fly when the project is built so we can keep auto-generated code out of git.

Thanks for working through all of these items, I think the patch is *really* close and a great addition.

danharvey commented:
Sure, added the maven generation and avro schema file.

ewencp commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
+  private Schema getReaderSchema(Schema writerSchema) {
+    Schema readerSchema = targetSchemaCache.get(writerSchema);
+    if (readerSchema == null) {
+      Class targetClass = SpecificData.get().getClass(writerSchema);
```
One thing I'm not clear on is how this handles different versions of the same schema. And the javadoc for this method doesn't give any details either. For example, if I have version 1 and version 2 of the schema registered, does getClass() only look at the `namespace` and `name` fields to look up the class, or does it actually compare the schema as well.

I think we need to think through the use cases to decide what the right behavior is. For example, if Avro only uses the name, should we also compare schemas and trigger an exception if they are incompatible? Or just leave that to the deserialization step which would presumably also fail? Would it be more useful to catch that early since we might be able to give a more meaningful error message.

danharvey replied:
From my point of view the use case for this is to convert data from other compatible schemas for a given subject to the schema that is compiled into the running code. We test and work against this compiled in schema so then as long as all schemas written into kafka are compatible then we know it will still work for new and old data.

The code for [`getClass()`](https://github.com/apache/avro/blob/trunk/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L150) looks at the namespace/name as you said from the schema and tries to lookup that class. So there will ever only be one reader schema per namespace/name, which could be for many different writer schemas.

For us subject is a 1:1 mapping with namespace/name so then the compatibility falls back to ensuring that all schemas in a subjected are **fully** compatible. We check this pre-merge for new schemas so we know they are consistent.

What we could do here to double check again is to check the readerSchema is compatible the first time it is found for a given namespace/name and raise an exception if it's not? Though that compatibility would be for every pair of schemas, so that would make it a bit more complicated. We would still only have to do it once per pair so probably better than eventually silently failing with a bad error and irrelevant message (which is what avro gives you I think).

ewencp commented:
Aside from the question about how exactly the class lookup works and whether we need to worry about compatibility issues there, this LGTM.

@junrao can you also review? You're much more familiar with the details of Avro and compatibility issues than I am.

junrao commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -19,16 +19,23 @@
 public abstract class AbstractKafkaAvroDeserializer extends AbstractKafkaAvroSerDe {
   private final DecoderFactory decoderFactory = DecoderFactory.get();
+  protected boolean useSpecificAvroReader = false;
+  private final Map<Schema, Schema> targetSchemaCache = new ConcurrentHashMap<Schema, Schema>();
```
We can just cache String => Schema where String can be obtained from writerSchema.getFullName() and schema is the reader schema, instead of writer schema => reader schema.

danharvey replied:
Yes, I've changed this now.

junrao commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
+      Class targetClass = SpecificData.get().getClass(writerSchema);
+      if (targetClass != null) {
+        try {
+          readerSchema = ((SpecificRecord) targetClass.newInstance()).getSchema();
```
We can also get the reader schema by calling the static method getClassSchema() on the class. Would that be simpler than instantiating an instance?

danharvey replied:
You can but then we'd have to use reflection and more code to actually call this. So this seemed a cleaner option. 

Given we're caching the result and only doing it once simpler code is probably better?

junrao commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -71,4 +78,35 @@ protected Object deserialize(byte[] payload) throws SerializationException {
+            "writers schema is not a SpecificRecord which is required to find the readers schema.");
+        } catch (IllegalAccessException e) {
+          throw new SerializationException(writerSchema.getFullName() + " specified by the " +
+            "writers schema is not a SpecificRecord which is required to find the readers schema.");
```
Could we distinguish the message string caused by InstantiationException and IllegalAccessException?

danharvey replied:
Yes, I've modified the messages so they are different.

ConfluentJenkins commented:
Can one of the admins verify this patch?

ewencp commented:
ok to test

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/1/
Test PASSed.

ewencp commented:
[clabot:check]

ewencp commented:
Ok, that didn't work quite as hoped. Until I sort out why this hook isn't working, @danharvey could you head over to http://clabot.confluent.io/cla and fill in the CLA? Should only take a minute, and I'd like to get this PR merged. Thanks!

danharvey commented:
Sure, just filled that in and submitted online.

ewencp commented:
Great, thanks for the patch!
